    <link name="base_link">
        <inertial>
        <mass value="${mass}" />
        <origin xyz="${cog}" rpy="0 0 0"/>
        <inertia ixx="0" ixy="0" ixz="0"
                iyy="0" iyz="0"
                izz="0" />
        </inertial>
    </link>
    <link name="world_link_x">
        <inertial>
        <mass value="0" />
        <origin xyz="0" rpy="0 0 0"/>
        <inertia ixx="0" ixy="0" ixz="0"
                iyy="0" iyz="0"
                izz="0" />
        </inertial>
    </link>
    <link name="world_link_y">
        <inertial>
        <mass value="0" />
        <origin xyz="0" rpy="0 0 0"/>
        <inertia ixx="0" ixy="0" ixz="0"
                iyy="0" iyz="0"
                izz="0" />
        </inertial>
    </link>
    <link name="world_link_z">
        <inertial>
        <mass value="0" />
        <origin xyz="0" rpy="0 0 0"/>
        <inertia ixx="0" ixy="0" ixz="0"
                iyy="0" iyz="0"
                izz="0" />
        </inertial>
    </link>
    <link name="world_link_roll">
        <inertial>
        <mass value="0" />
        <origin xyz="0" rpy="0 0 0"/>
        <inertia ixx="0" ixy="0" ixz="0"
                iyy="0" iyz="0"
                izz="0" />
        </inertial>
    </link>
    <link name="world_link_pitch">
        <inertial>
        <mass value="0" />
        <origin xyz="0" rpy="0 0 0"/>
        <inertia ixx="0" ixy="0" ixz="0"
                iyy="0" iyz="0"
                izz="0" />
        </inertial>
    </link>

      <!-- Rigid body description of the base link -->
      <!-- <link name="${namespace}/base_link"> -->
      <link name="world_link_yaw">
        <!--
          Be careful to setup the coefficients for the inertial tensor,
          otherwise your model will become unstable on Gazebo
        -->
        <inertial>
          <mass value="0" />
          <origin xyz="0" rpy="0 0 0"/>
          <inertia ixx="0" ixy="0" ixz="0"
                   iyy="0" iyz="0"
                   izz="0" />
        </inertial>

        <!-- <visual>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <geometry> -->
            <!-- <mesh filename="${visual_mesh_file}" scale="1 1 1" />  -->
            <!-- <box size="1 1 1" />
            <color rgb="blue" />
          </geometry>
        </visual>

        <collision>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <geometry>
            <box size="1 1 1" />
          </geometry>
        </collision> -->
      </link>

      <!-- robot name="OSU vLBV300 Volturnus" -->
      <xacro:property name="pi" value="3.1515926535897931" />
      <xacro:property name="deg2rad" value="${pi/180}" />

      <!-- Without trig, angles are a pain in the arse. Beam angle 30 -->
      <xacro:property name="dvl_beam_angle" value="${30/180*pi}" />
      <xacro:property name="dvl_pitch" value="0.361367" />
      <xacro:property name="dvl_yaw" value="0.387597" />


      <material name="green_v">
        <color rgba="0.5 0.9 0.3 1"/>
      </material>
      <material name="black_v">
        <color rgba="0 0 0 1"/>
      </material>
      <material name="red_v">
        <color rgba="1 0 0 1"/>
      </material>
      <material name="blue_v">
        <color rgba="0.2 0.4 0.6 1" />
      </material>
      <material name="grey_v">
        <color rgba=".3 .3 .3 1"/>
      </material>
      <material name="orange_v">
        <color rgba="1 .4 0 1" />
      </material>

      <link name="vehicle_base_link">
        <visual>
          <geometry>
            <box size="0.5 0.45 0.15"/>
          </geometry>
          <material name="green_v"/>
        </visual>
      </link>

      <link name="front_bumper">
        <visual>
          <geometry>
            <cylinder length="0.45" radius="0.075"/>
          </geometry>
          <material name="green_v"/>
        </visual>
      </link>
      <joint name="fbump_link" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="front_bumper"/>
        <origin rpy="${pi/2} 0 0" xyz="0.25 0 0"/>
      </joint>

      <link name="rear_bumper">
        <visual>
          <geometry>
            <cylinder length="0.45" radius="0.075"/>
          </geometry>
          <material name="green_v"/>
        </visual>
      </link>
      <joint name="rbump_link" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="rear_bumper"/>
        <origin rpy="${pi/2} 0 0" xyz="-0.25 0 0"/>
      </joint>

      <link name="l_arrow">
        <visual>
          <geometry>
            <box size="0.05 0.318 0.01"/>
          </geometry>
          <material name="black_v"/>
        </visual>
      </link>
      <joint name="l_arrow_link" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="l_arrow"/>
        <origin rpy="0 0 ${pi/4}" xyz="0.05 0.095 0.075"/>
      </joint>

      <link name="r_arrow">
        <visual>
          <geometry>
            <box size="0.05 0.318 0.01"/>
          </geometry>
          <material name="black_v"/>
        </visual>
      </link>
      <joint name="r_arrow_link" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="r_arrow"/>
        <origin rpy="0 0 ${-pi/4}" xyz="0.05 -0.095 0.075"/>
      </joint>


      <link name="sf_fan">
        <visual>
          <geometry>
            <cylinder length="0.04" radius="0.06"/>
          </geometry>
          <material name="blue_v"/>
        </visual>
      </link>
      <joint name="base_to_sf_fan" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="sf_fan"/>
        <origin rpy="0 ${pi/2} 0.785" xyz="0.22 -0.12 -0.14"/>
      </joint>
      <link name="sf_thruster">
        <visual>
          <geometry>
            <cylinder length="0.1" radius="0.02"/>
          </geometry>
          <material name="black_v"/>
        </visual>
      </link>
      <joint name="sf_fan_thruster_joint" type="fixed">
        <parent link="sf_fan"/>
        <child link="sf_thruster"/>
        <origin xyz="0 0 -0.06"/>
      </joint>

      <link name="pf_fan">
        <visual>
          <geometry>
            <cylinder length="0.04" radius="0.06"/>
          </geometry>
          <material name="blue_v"/>
        </visual>
      </link>
      <joint name="base_to_pf_fan" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="pf_fan"/>
        <origin rpy="0 ${pi/2} -0.785" xyz="0.22 0.12 -0.14"/>
      </joint>
      <link name="pf_thruster">
        <visual>
          <geometry>
            <cylinder length="0.1" radius="0.02"/>
          </geometry>
          <material name="black_v"/>
        </visual>
      </link>
      <joint name="pf_fan_thruster_joint" type="fixed">
        <parent link="pf_fan"/>
        <child link="pf_thruster"/>
        <origin xyz="0 0 -0.06"/>
      </joint>

      <link name="sa_fan">
        <visual>
          <geometry>
            <cylinder length="0.04" radius="0.06"/>
          </geometry>
          <material name="blue_v"/>
        </visual>
      </link>
      <joint name="base_to_sa_fan" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="sa_fan"/>
        <origin rpy="0 ${-pi/2} -0.785" xyz="-0.22 -0.12 -0.14"/>
      </joint>
      <link name="sa_thruster">
        <visual>
          <geometry>
            <cylinder length="0.1" radius="0.02"/>
          </geometry>
          <material name="black_v"/>
        </visual>
      </link>
      <joint name="sa_fan_thruster_joint" type="fixed">
        <parent link="sa_fan"/>
        <child link="sa_thruster"/>
        <origin xyz="0 0 -0.06"/>
      </joint>

      <link name="pa_fan">
        <visual>
          <geometry>
            <cylinder length="0.04" radius="0.06"/>
          </geometry>
          <material name="blue_v"/>
        </visual>
      </link>
      <joint name="base_to_pa_fan" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="pa_fan"/>
        <origin rpy="0 -${pi/2} 0.785" xyz="-0.22 0.12 -0.14"/>
      </joint>
      <link name="pa_thruster">
        <visual>
          <geometry>
            <cylinder length="0.1" radius="0.02"/>
          </geometry>
          <material name="black_v"/>
        </visual>
      </link>
      <joint name="pa_fan_thruster_joint" type="fixed">
        <parent link="pa_fan"/>
        <child link="pa_thruster"/>
        <origin xyz="0 0 -0.06"/>
      </joint>

      <link name="seabotix_bottle">
        <visual>
          <geometry>
            <cylinder length="0.26" radius="0.05"/>
          </geometry>
          <material name="black_v"/>
        </visual>
      </link>
      <joint name="sbottle_link" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="seabotix_bottle"/>
        <origin rpy="0 ${pi/2} 0" xyz="-0.05 0 -0.18"/>
      </joint>
<!-- 
      <link name="greensea">
        <visual>
          <geometry>
            <cylinder length="0.2" radius="0.04"/>
          </geometry>
          <origin rpy="0 ${pi/2} 0"/>
          <material name="grey_v"/>
        </visual>
      </link>

      <joint name="gbottle_link" type="fixed">
        <parent link="greensea"/>
        <child link="vehicle_base_link"/>
        <origin rpy="-${pi} 0 0" xyz="0 0 -0.27"/>
      </joint> -->

      <link name="dvl">
        <visual>
          <geometry>
            <cylinder length="0.25" radius="0.06"/>
          </geometry>
          <material name="grey_v"/>
          <origin rpy="0 ${pi/2} 0" xyz="-0.125 0 0"/>
        </visual>
      </link>
      <joint name="dvl_link" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="dvl"/>
        <origin rpy="0 0 ${pi}" xyz="-0.23 -0.1 -0.29"/>
      </joint>

      <link name="dvl_head">
        <visual>
          <geometry>
            <sphere radius="0.06"/>
          </geometry>
          <material name="orange_v"/>
        </visual>
      </link>
      <joint name="dvl_head_link" type="fixed">
        <parent link="dvl"/>
        <child link="dvl_head"/>
        <origin rpy="0 0 ${-45*deg2rad}" xyz="0 0 0"/>
      </joint>

      <link name="dvl_altitude"/>
      <joint name="dvl_altitude_link" type="fixed">
        <parent link="dvl_head"/>
        <child link="dvl_altitude"/>
        <origin rpy="0 ${pi/2} 0" xyz="0 0 0"/>
      </joint>

      <link name="dvl_beam1">
        <visual>
          <geometry>
            <cylinder length="0.07" radius="0.015"/>
          </geometry>
          <origin rpy="0 ${pi/2} 0" xyz="0.035 0 0"/>
          <material name="blue_v"/>
        </visual>
      </link>
      <joint name="dvl_beam1_link" type="fixed">
        <parent link="dvl_head"/>
        <child link="dvl_beam1"/>
        <origin rpy="0 ${pi/2-dvl_beam_angle} ${pi/2}" xyz="0 0 0"/>
      </joint>

      <link name="dvl_beam2">
        <visual>
          <geometry>
            <cylinder length="0.07" radius="0.015"/>
          </geometry>
          <origin rpy="0 ${pi/2} 0" xyz="0.035 0 0"/>
          <material name="blue_v"/>
        </visual>
      </link>
      <joint name="dvl_beam2_link" type="fixed">
        <parent link="dvl_head"/>
        <child link="dvl_beam2"/>
        <origin rpy="0 ${pi/2-dvl_beam_angle} -${pi/2}" xyz="0 0 0"/>
      </joint>

      <link name="dvl_beam3">
        <visual>
          <geometry>
            <cylinder length="0.07" radius="0.015"/>
          </geometry>
          <origin rpy="0 ${pi/2} 0" xyz="0.035 0 0"/>
          <material name="blue_v"/>
        </visual>
      </link>
      <joint name="dvl_beam3_link" type="fixed">
        <parent link="dvl_head"/>
        <child link="dvl_beam3"/>
        <origin rpy="0 ${pi/2-dvl_beam_angle} 0" xyz="0 0 0"/>
      </joint>

      <link name="dvl_beam4">
        <visual>
          <geometry>
            <cylinder length="0.07" radius="0.015"/>
          </geometry>
          <origin rpy="0 ${pi/2} 0" xyz="0.035 0 0"/>
          <material name="blue_v"/>
        </visual>
      </link>
      <joint name="dvl_beam4_link" type="fixed">
        <parent link="dvl_head"/>
        <child link="dvl_beam4"/>
        <origin rpy="0 ${pi/2-dvl_beam_angle} ${pi}" xyz="0 0 0"/>
      </joint>

      <link name="sonar">
        <visual>
          <geometry>
            <box size="0.22 0.13 0.1"/>
          </geometry>
          <material name="orange_v"/>
        </visual>
      </link>
      <joint name="sonar_link" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="sonar"/>
        <origin xyz="0.1 0.12 -0.28"/>
      </joint>

      <link name="gripper_pole">
        <visual>
          <geometry>
            <cylinder length="0.38" radius="0.015"/>
          </geometry>
          <origin rpy="0 ${pi/2} 0" xyz="0.19 0 0"/>
          <material name="black_v"/>
        </visual>
      </link>
      <joint name="gripper_pole_link" type="fixed">
        <parent link="vehicle_base_link"/>
        <child link="gripper_pole"/>
        <origin rpy="0 0.1 0.3" xyz="0.15 -0.15 -0.2"/>
      </joint>



    <joint name="x_displacement" type="prismatic">
      <parent link="base_link"/>
      <child link="world_link_x"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <axis xyz="1 0 0"/>
      <limit effort="100" velocity="0.5" lower="-10" upper="10" /> <!-- Assume we are operating in 20x20x20m cube-->
    </joint>

    <joint name="y_displacement" type="prismatic">
      <parent link="world_link_x"/>
      <child link="world_link_y"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <axis xyz="0 1 0"/>
      <limit effort="100" velocity="0.5" lower="-10" upper="10" /> <!-- Assume we are operating in 20x20x20m cube-->
    </joint>

    <joint name="z_displacement" type="prismatic">
      <parent link="world_link_y"/>
      <child link="world_link_z"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <axis xyz="0 0 1"/>
      <limit effort="100" velocity="0.5" lower="-10" upper="10" /> <!-- Assume we are operating in 20x20x20m cube-->
    </joint>

    <joint name="roll_displacement" type="revolute">
      <parent link="world_link_z"/>
      <child link="world_link_roll"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <axis xyz="1 0 0"/>
      <limit effort="100" velocity="${pi/10}" lower="${-pi/2}" upper="${pi/2}" /> <!-- Assume we can roll by pi/2-->
    </joint>

    <joint name="pitch_displacement" type="revolute">
      <parent link="world_link_roll"/>
      <child link="world_link_pitch"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <axis xyz="0 1 0"/>
      <limit effort="100" velocity="${pi/10}" lower="${-pi/2}" upper="${pi/2}" /> <!-- Assume we can pitch by pi/2-->
    </joint>

    <joint name="yaw_displacement" type="continuous">
      <parent link="world_link_pitch"/>
      <child link="world_link_yaw"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <axis xyz="0 0 1"/>
      <limit effort="100" velocity="${pi/10}"/> <!-- Assume we have no limit on yaw-->
    </joint>

    <!-- Add a link attaching the vehicle to the world base link -->
    <joint name="joint_vehicle_base_world" type="fixed">
      <!-- <parent link="${namespace}/base_link"/> -->
      <parent link="world_link_yaw"/>
      <child link="vehicle_base_link"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </joint>